Update and expand docs and code comments related to Plutip testing (#1504)
Plutonomicon/cardano-transaction-lib: doc/plutip-testing.md

- [Architecture](#architecture)
- [Testing contracts](#testing-contracts)
  - [Testing with Mote overview](#testing-with-mote-overview)

>   - [Testing in Aff context overview](#testing-in-aff-context-overview)

Reviewer: Let's organize it this way:
- Testing with Mote
  - Overview
  - Using Mote testing interface
  - Internal implementation overview
- Testing in Aff context
- Writing checks in tests

Author: restructed it like this

- [`ogmios`](https://ogmios.dev/)
- [`kupo`](https://cardanosolutions.github.io/kupo/)

> All of these are provided by CTL's `overlays.runtime` (and are provided in CTL's own `devShell`). You **must** use the `runtime` overlay or otherwise make the services available in your package set (e.g. by defining them within your own `overlays` when instantiating `nixpkgs`) as `purescriptProject.runPlutipTest` expects all of them; an example of using CTL's overlays is in a [`ctl-scaffold` template](../templates/ctl-scaffold/flake.nix#L35).

Reviewer: ```suggestion
All of these are provided by CTL's `overlays.runtime` (and are provided in CTL's own `devShell`). You **must** use the `runtime` overlay or otherwise make the services available in your package set (e.g. by defining them within your own `overlays` when instantiating `nixpkgs`) as `purescriptProject.runPlutipTest` expects all of them; an example of using CTL's overlays is in the [`ctl-scaffold` template](../templates/ctl-scaffold/flake.nix#L35).
```

The services are NOT run by `docker-compose` (via `arion`) as is the case with `launchCtlRuntime`: instead they are started and stopped on each CTL `ContractTest` execution by CTL itself.

If you have based your project on the [`ctl-scaffold` template](../templates/ctl-scaffold) then you have two options to run Plutip tests:
1. `nix develop` followed by `npm run test` (recommended for development)
2. `nix run .#checks.x86_64-linux.ctl-scaffold-plutip-test`
   * where you'd usually replace `x86_64-linux` with the system you run tests on

>    * and `ctl-scaffold-plutip-test` with the name of the plutip test package for your project;

Reviewer: ```suggestion
   * and `ctl-scaffold-plutip-test` with the name of the plutip test derivation for your project;
```

>    * note that compilation of your project via Nix will fail in case there are any warnings, so you'd need to either fix or temporarily disable them.

Reviewer: Why this warning?

Author: Mainly because `npm run test` will run plutip tests even if there are warnings, so compared to to it `nix run` refusing to run tests might come as a surprise if you just want to quickly run some tests. Compilation with `nix run` also starts from scratch and is much slower, so this part of doc manages some expectations people might have.

>    * note that compilation of your project via Nix will fail in case there are any warnings, so you'd need to either fix or temporarily disable them.

Reviewer: ```suggestion
   * note that building of your project via Nix will fail in case there are any PureScript compile-time warnings.
```

Reviewer: The warnings can't be disabled

## Testing contracts

CTL can help you test the offchain `Contract`s from your project (and consequently the interaction of onchain and offchain code) by spinning up a disposable private testnet via Plutip and making all your `Contract`s interact with it.

> There are two approaches to writing such tests.

Reviewer: Please use titles (testing with mote/testing in Aff context). Mote testing should come first, because it's a go-to method.

After the contract execution the Plutip cluster is terminated.
You can either call it directly from your test's main or use any library for grouping and describing tests which support effects in the test body, like Mote.

[Mote](https://github.com/garyb/purescript-mote) is a DSL for defining and grouping tests (plus other quality of life features, e.g. skipping marked tests).

[Second](#testing-with-mote) (and more widely used) approach is to first build a tree of tests (in CTL's case a tree of `ContractTest` types -- basically a function from some distribution of funds to a `Contract a`) via Mote and then use the `Contract.Test.Plutip.testPlutipContracts` function to execute them.
This allows to set up a Plutip cluster only once per top-level groups and tests passed to the `testPlutipContracts` and then use it in many independent tests.
The function will interpret a `MoteT` (effectful test tree) into `Aff`, which you can then actually run.

The [`ctl-scaffold` template](../templates/ctl-scaffold) provides a simple `Mote`-based example.

CTL will run contracts in your test bodies and will print errors for any failed tests.

> Only test body failures are checked and this works fine if you want to make sure your `Contract`s execute without errors; if you want to add more precise checks (like checking that particular token is now at some address, that some exact amount was transferred, etc.) then you need to either write these checks manually in a `Contract` monad and then throw errors, or (preferably) use the [assertions library](./test-utils.md).

Reviewer: > either

It's not either. It's either or both. Re-phrase it starting with "there are two options available..."

Author: Fixed

The communication with Plutip happens via the `plutip-server`'s HTTP interface, which allows to start or stop a cluster.

> [`plutip-server`](../plutip-server) allows only once active cluster at a time, but nothing stops you from setting up multiple CTL environments and multiple `plutip-server`s by running tests in separate fibers and thus using multiple Plutip clusters simultaneously.

Reviewer: This is not true. It uses a set of pre-defined ports, so parallel clusters will fail to start unless the ports are explicitly made different.

Author: Internally Plutip uses `cardano-wallet` function `randomUnusedTCPPorts` to get a list of ports later assigned to nodes: https://github.com/input-output-hk/cardano-wallet/blob/af82118b5cd5addc60c68dc4fdaf59cb1d228be7/lib/wallet/src/Cardano/Wallet/Network/Ports.hs#L119

So you can run parallel clusters, but nodes in different clusters can be assigned the same port due to race condition, and then there'll be another race condition where one of the nodes will fail to start.
I'll rephrase and mention this in the doc.

Author: Ah, I see, yes this is a problem then, will mention this

Author: Fixed

There are two approaches to writing such tests.

### Testing with Mote overview

[Mote](https://github.com/garyb/purescript-mote) is a DSL for defining and grouping tests (plus other quality of life features, e.g. skipping marked tests).

First (and more widely used) approach is to first build a tree of tests (in CTL's case a tree of `ContractTest` types -- basically a function from some distribution of funds to a `Contract a`) via Mote and then use the `Contract.Test.Plutip.testPlutipContracts` function to execute them.
This allows to set up a Plutip cluster only once per top-level groups and tests passed to the `testPlutipContracts` and then use it in many independent tests.
The function will interpret a `MoteT` (effectful test tree) into `Aff`, which you can then actually run.

The [`ctl-scaffold` template](../templates/ctl-scaffold) provides a simple `Mote`-based example.

More info and usage examples are [here](#testing-with-mote).

### Testing in Aff context overview

> [First](#testing-in-aff-context) is to use the `Contract.Test.Plutip.runPlutipContract` function, which takes a single `Contract`, launches a Plutip cluster and executes the passed contract.

Reviewer: This is duplicate. "First" is used above already in a similar paragraph.

Author: Fixed

This function runs in `Aff`; it will also throw an exception should contract fail for any reason.
After the contract execution the Plutip cluster is terminated.
You can either call it directly from your test's main or use any library for grouping and describing tests which support effects in the test body, like Mote.

### Writing checks in tests

CTL will run contracts in your test bodies and will print errors for any failed tests.

> Only test body failures are checked and this works fine if you want to make sure your `Contract`s execute without errors; if you want to add more precise checks (like checking that particular token is now at some address, that some exact amount was transferred, etc.) then you need to write these checks manually in a `Contract` monad (and preferably to utilize the [assertions library](./test-utils.md)) and then throw errors.

Reviewer: This paragraph can be simplified a lot. Just make a note that assertion library exists and can be used for complex checks. Users already know that exceptions result in a failure.

Author: Condensed it to a simple sentence

### Note on running parallel clusters

The communication with Plutip happens via the `plutip-server`'s HTTP interface, which allows to start or stop a cluster.

> [`plutip-server`](../plutip-server) allows only once active cluster at a time, but nothing stops you from setting up multiple CTL environments and multiple `plutip-server`s by running tests in separate fibers and thus using multiple Plutip clusters simultaneously. One caveat is that nodes in different clusters might get assigned the same port (see [this](https://github.com/mlabs-haskell/plutip/blob/master/README.md#note-on-running-multiple-clusters) Plutip doc) and then race to use it, which will result in one cluster starting fine and another repeatedly failing. The way to deal with this is to start another environment and try again.

Reviewer: This is still not true.
Not only cardano-node will have port conflicts, but all runtime services (Ogmios and Kupo).

> The way to deal with this is to start another environment and try again.

It will not work. The only way to deal with port conflicts is to use different ports in the second plutip config.

Author: I think github showed an outdated version here, I've fixed it and moved part of the original description inside a comment

  -> Aff a
```

> `distr` is a specification of how many wallets and with how much funds should be created. It should either be a `Unit` (for no wallets), nested tuples containing `Array BigInt` or an `Array` of `Array BigInt`, where each element of the `Array BigInt` specifies an UTxO amount in Lovelaces (0.000001 Ada).

Reviewer: ```suggestion
`distr` is a specification of how many wallets and with how much funds should be created. It should either be a `Unit` (for no wallets), nested tuples containing `Array BigInt` or an `Array (Array BigInt)`, where each element of the inner array specifies an UTxO amount in Lovelaces (0.000001 Ada).
```

In most cases at least two UTxOs per wallet are needed (one of which will be used as collateral, so it should exceed `5_000_000` Lovelace).

> Internally `runPlutipContract` runs a contract in an `Aff.bracket`, which creates Plutip cluster during the setup and terminates it during the teardown or in the case of an exception.

Reviewer: ```suggestion
Internally `runPlutipContract` runs a contract in an `Aff.bracket`, which creates a Plutip cluster on setup and terminates it during the shutdown or in case of an exception.
```

> Logs will be printed in case of error.

Reviewer: ```suggestion
Logs will be printed in case of an error.
```

### Testing with Mote

`Contract.Test.Plutip.testPlutipContracts` type is defined as follows (after expansion of the CTL's `TestPlanM` type synonym):
```purescript
type TestPlanM :: Type -> Type -> Type
type TestPlanM test a = MoteT Aff test Aff a

testPlutipContracts
  :: PlutipConfig
  -> MoteT Aff ContractTest Aff Unit
  -> MoteT Aff (Aff Unit) Aff Unit

-- Recall that `MoteT` has three type variables
newtype MoteT bracket test m a
```
where

> * `bracket :: Type -> Type` is where brackets will be run (before/setup is `bracket r` and after/teardown is of type `r -> bracket Unit`),

Reviewer: ```suggestion
* `bracket :: Type -> Type` is where brackets will be run (before/setup is `bracket r` and after/shutdown is of type `r -> bracket Unit`),
```

   * in our case it's `Aff` and is where the CTL environment and Plutip cluster setup will happen,
   * also environment setup and Plutip startup and teardown will happen once per each top-level test or group inside the `testPlutipContracts` call,
   * so wrap your tests or groups in a single group if you want for the cluster to start only once,
* `test :: Type` is a type of tests themselves,
   * in our case it's [`ContractTest`](../src/Internal/Test/ContractTest.purs), which in a nutshell describes a function from some wallet UTxO distribution to a `Contract r`
   * wallet UTxO distribution is the one that you need to pattern-match on when writing tests
* `m :: Type -> Type` is a monad where effects during the construction of the test suite can be performed,
   * here we use `Aff` again

> * `a :: Type` is a result of the test suite, we use `Unit` here.

Reviewer: This whole section should me de-prioritized. We must first tell the user how to use `withWallets`, and only then dive into the techinical detail of an internal type. Move it to a separate sub-section and state at the beginning of the paragraph that this is an explanation how Mote and our intergation with it works *internally*

Author: Moved it to a separate section, combined with other internal notes

      ...
```

> To define tests suites you can use `test`, group them with `group` and also wrap tests or groups with `bracket` to execute custom actions before and after tests/groups that are inside the bracket.

Reviewer: Add a new section "Using Mote testing interface"

Author: Fixed

See the comments in the [`Ctl.Internal.Plutip.Server` module](../src/Internal/Plutip/Server.purs) for more info (relevant ones are in `execDistribution` and `testPlutipContracts` functions).

In complicated protocols you might want to execute some `Contract`s in one test and then execute other `Contract`s which depend on some wallet-dependent state set up by the first batch of contracts, e.g. some authorization token is present at some wallet.
Keeping these steps in separate sequential tests allows to pinpoint where things failed much easier, but currently CTL uses separate wallets for each test without an easy way to refer to wallets in other tests, so you have to call first batch of contracts again to replicate the state of the wallets, which in turn might fail or mess up your protocol, because the chain state is shared between tests for each top-level group.
There's a patch to CTL you can adapt (and even better -- make a PR) if you need to share wallets between tests right now, see the [limitations](#limitations) doc for more info.
This functionality will probably be added to CTL later.

### Note on SIGINT

> <!-- TODO: clarify which node is meant here -- node.js or cardano-node? -->

Reviewer: nodejs

> <!-- TODO: why does the CTL project template restore the default behavior? -->

Reviewer: Because it provides ability to Ctrl+C the test suite properly

Internally `runPlutipContract` runs a contract in an `Aff.bracket`, which creates a Plutip cluster on setup and terminates it during the shutdown or in case of an exception.
Logs will be printed in case of an error.

### Note on SIGINT

> Due to `testPlutipContracts`/`runPlutipContract` adding listeners to the SIGINT signal, Node.js's default behaviour of exiting on that signal no longer occurs. This was done to add cleanup handlers and let them run in parallel instead of exiting eagerly, which is possible when running multiple clusters in parallel. To restore the exit behaviour, we provide helpers to cancel an `Aff` fiber and set the exit code, to let Node.js shut down gracefully when no more events are to be processed.

Reviewer: ```suggestion
Due to `testPlutipContracts`/`runPlutipContract` adding listeners to the SIGINT  IPC signal, Node.js's default behaviour of exiting on CTRL+C no longer occurs. This was done to let cluster cleanup handlers run asynchronously. To restore the usual exit-by-CTRL+C behaviour, we provide helpers to cancel an `Aff` fiber and set the exit code, to let Node.js shut down gracefully when no more events are to be processed.
```

}
```

The usual approach is to put `projectname-plutip-test` in the `checks` attribute of your project's `flake.nix`.

> See a [`ctl-scaffold` template](../templates/ctl-scaffold/flake.nix) for an example.

Reviewer: ```suggestion
This is done by default in the [`ctl-scaffold` template](../templates/ctl-scaffold/flake.nix).
```

- `maxTxSize` (in bytes) allows to stress-test protocols with more restrictive transaction size limits.
- `raiseExUnitsToMax` allows to bypass execution units limit (useful when compiling the contract with tracing in development and without it in production).

### Current limitations

> * Non-default value of `epochSize` (current default is 80) break staking rewards - see [this issue](https://github.com/mlabs-haskell/plutip/issues/149) for more info. `slotLength` can be changed without any problems.

Reviewer: ```suggestion
* Non-default values of `epochSize` (current default is 80) break staking rewards - see [this issue](https://github.com/mlabs-haskell/plutip/issues/149) for more info. `slotLength` can be changed without any problems.
```

Note that CTL re-distributes tADA from payment key-only ("enterprise") addresses to base addresses, which requires a few transactions before the test can be run. These transactions happen on the CTL side, because Plutip can currently handle only enterprise addreses (see [this issue](https://github.com/mlabs-haskell/plutip/issues/103)).

## Limitations
* See the `epochSize` configuration option problem [here](#current-limitations).
* Currently there's no way to share wallets between separate tests (which is useful for complex protocols). You can adapt [this PR](https://github.com/IndigoProtocol/cardano-transaction-lib/pull/1) (needs to be updated for the newer versions of CTL, likely won't need too many changes) if you need it now (and even better -- make a PR to CTL).

> * If you've used a [`plutus-simple-model`](https://github.com/mlabs-haskell/plutus-simple-model) library then you might know that it allows time travel in tests, which can be very useful for testing vesting schedules, etc. Testing with Plutip doesn't allow this, as it's running a real network, so you'd have to really wait, say 1 year, in order to test interaction with onchain code which checks that much time has passed.

Reviewer: Suggest making the offchain logic parametrizeable by a time multiplier

## See also

> - To actually write the test bodies, [assertions library](./test-utils.md) can be useful.
> - Check out the [`ContractTestUtils`](../examples/ContractTestUtils.purs) example on assertions library usage.

Reviewer: ```suggestion
- To actually write the test bodies, [assertions library](./test-utils.md) can be useful [(usage example)](../examples/ContractTestUtils.purs).
```

## Limitations
* See the `epochSize` configuration option problem [here](#current-limitations).
* Currently there's no way to share wallets between separate tests (which is useful for complex protocols). You can adapt [this PR](https://github.com/IndigoProtocol/cardano-transaction-lib/pull/1) (needs to be updated for the newer versions of CTL, likely won't need too many changes) if you need it now (and even better -- make a PR to CTL).

> * If you've used a [`plutus-simple-model`](https://github.com/mlabs-haskell/plutus-simple-model) library then you might know that it allows time travel in tests, which can be very useful for testing vesting schedules, etc. Testing with Plutip doesn't allow this, as it's running a real network, so you'd have to really wait, say 1 year, in order to test interaction with onchain code which checks that much time has passed. A way around this problem can be to parametrize onchain logic by a time multiplier (and use a small one for tests).

Reviewer: ```suggestion
* If you've used the [`plutus-simple-model`](https://github.com/mlabs-haskell/plutus-simple-model) library then you might know that it allows to time travel in tests, which can be very useful for testing vesting schedules, etc. Testing with Plutip doesn't allow this, as it's running a real network. A way around this problem can be to parametrize onchain logic by a time multiplier (and use a small one for tests).
```
